### PR TITLE
12 feature 0 1 writing out the entities and interfaces user entities

### PR DESCRIPTION
**Pull request by xuyangan**
The User Entities have been added. There are minimal methods and we will add more as we go through the use cases. But there should have the necessary attributes, let me know if I am missing something. Also, I didn't complete the test cases because the other entities are missing. I'll get to them when the shared code is a little bit more complete.

### CODE REVIEW

File: src/main/Entities/Researcher.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package Entities;
+
+import java.util.ArrayList;
+
+public class Researcher extends User {
+
+    /**
+     * A list of Study objects that the this researcher belongs to (works in).
+     */
+    private ArrayList<Study> listStudies = new ArrayList<Study>();
```
Comment (j-y26) [inline src/main/Entities/Researcher.java:10]: Make listStudies a "List". Variable types should be as abstract as possible.

File: src/main/Entities/Researcher.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    public Researcher(String username, String name) {
+        super(username, name);
+
+    }
+
+    /**
+     * @return the list of Study objects this Researcher belongs to, i.e. works in
+     */
+    public ArrayList<Study> getListStudies() {
```
Comment (j-y26) [inline src/main/Entities/Researcher.java:24]: Similarly the return type should be List<Study>

File: src/main/Entities/Researcher.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package Entities;
+
+import java.util.ArrayList;
```
Comment (j-y26) [inline src/main/Entities/Researcher.java:3]: import List

File: src/main/Entities/Researcher.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    /**
+     * @return the list of Study objects this Researcher belongs to, i.e. works in
+     */
+    public ArrayList<Study> getListStudies() {
+        return listStudies;
+    }
+
+    public boolean addToListStudies(Study study){
+        listStudies.add(study);
+        return true;
```
Comment (j-y26) [inline src/main/Entities/Researcher.java:30]: Should also check whether the study to be added already exist in listStudies.

File: src/main/Entities/Researcher.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    public boolean removeFromListStudies(Study study){
+        if (listStudies.contains(study)){
```
Comment (j-y26) [inline src/main/Entities/Researcher.java:34]: Make sure contains, remove, add are methods defined by the abstract type List.

File: src/main/Entities/Participant.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package Entities;
```
Comment (j-y26) [inline src/main/Entities/Participant.java:1]: The package name should be in all lowercase: change to "entities" instead. I think this is how the Gradle project is setup.

**Review by j-y26 — CHANGES_REQUESTED** [review]

**Review by j-y26 — DISMISSED** [review]

File: src/main/entities/Researcher.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+    /**
+     * @return the list of Study objects this Researcher belongs to, i.e. works in.
+     */
+    public List<Study> getListStudies() {
+        return listStudies;
+    }
+
+    /**
+     * @param study
```
Comment (chelkisea) [inline src/main/entities/Researcher.java:29]:
```suggestion
     * @param study a Study object that this Researcher wants to check if it is contained in listStudies.

```

File: src/main/entities/Researcher.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+     * @return true if this Researcher is part of study.
+     */
+    public boolean listStudiesContains(Study study) {
+        return listStudies.contains(study);
+    }
+
+    /**
+     * Enroll this researcher to study
+     *
+     * @param study
```
Comment (chelkisea) [inline src/main/entities/Researcher.java:39]:
```suggestion
     * @param study a Study object that this Researcher wants to be enrolled in.

```

File: src/main/entities/Researcher.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        }
+        listStudies.add(study);
+        return true;
+    }
+
+    /**
+     * Removes the a Study from this researcher's list of Studies. Either because the study was close, or the Researcher
+     * is no longer part of the Study.
+     *
+     * @param study
```
Comment (chelkisea) [inline src/main/entities/Researcher.java:54]:
```suggestion
     * @param study a Study object that this Researcher wants to leave.

```

File: src/main/entities/UserFactory.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package entities;
+
+public class UserFactory implements UserFactoryInterface{
+
+    /**
+     *
```
Comment (chelkisea) [inline src/main/entities/UserFactory.java:6]:
```suggestion
     *
     *Precondition: the type of user must be either "Participant" or "Researcher"

```

File: src/main/entities/Researcher.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+     * @return
+     */
+    public boolean removeFromListStudies(Study study) {
+        if (this.listStudiesContains(study)) {
+            listStudies.remove(study);
+            return true;
+        }
+        return false;
+    }
+
```
Comment (chelkisea) [inline src/main/entities/Researcher.java:64]: If we remove/add a researcher from the study, shouldn't we update the attribute 'researchers' in class Study as well?

**Review by chelkisea — CHANGES_REQUESTED** [review]
please check review comments

File: src/main/entities/Researcher.java (hunk 11 above)
Comment (xuyangan) [reply src/main/entities/Researcher.java:64]: Yes! but I believe we have a Use Case that calls on this method when a researcher is removed. That same Use Case will have to call on the Study object this researcher belongs to, to remove this researcher from its attribute (i.e. the Study class should have a removeResearcher method. So a researcher object doesn't directly call a method of a study object.

Comment (chelkisea) [reply src/main/entities/Researcher.java:67]: got it,thanks!

**Review by chelkisea — APPROVED** [review]

**Review by j-y26 — APPROVED** [review]

**Review by j-y26 — APPROVED** [review]

**Review by Jerryfish0807 — DISMISSED** [review]

**Review by Tedward765 — DISMISSED** [review]
not even much of an error but there is a missing newline and the end of the code (not sure if java still runs if its missing)

**Review by j-y26 — APPROVED** [review]

**Review by j-y26 — APPROVED** [review]

**Review by j-y26 — APPROVED** [review]

**Review by Jerryfish0807 — DISMISSED** [review]

**Review by j-y26 — APPROVED** [review]

**Review by j-y26 — APPROVED** [review]

**Review by CoqRooster — APPROVED** [review]
Well done!

**Review by xustell — APPROVED** [review]
wow

**Pull request merged**